veps/sig-compute/decentralized-migration/decentralized-migration-vep.md

```diff
@@ -0,0 +1,308 @@
+
+The actual migration channel is proxied through the `virt-handler` pods that are running on the source and target nodes. Any migration state information is passed through the `VirtualMachineInstance` migrationStatus. This is because `virt-handler` pods can only read/modify `VirtualMachineInstances`.
+
+The reason virt-launcher pods do not communicate directly with each other, is that it is not guaranteed that virt-launchers will have network connectivity. The virt-handler pods also have access to a migration network if one is defined.
+
+The migration channel in figure 1 is either the pod network, or a specially configured migration network.
+
+Figure 1 describes the current interaction between various components:
+![figure 1](migration_current_state.png)
+
+#### Proposed state
+To start a live migration one now has to create two `VirtualMachineInstanceMigration` resources. A `target` resource, and a `source` resource. Additionally in order to migrate a special `VirtualMachine` resource has to be created on the target which is copy of the source `VirtualMachine` but it has a special `runStrategy` indicating it is the receiving `VirtualMachine`. The virt-controller will create a special receiving `VirtualMachineInstance` that can be referenced by the `target` `VirtualMachineInstanceMigration`.
+
+Additionally some other resources may be required if they are referenced by the source `VirtualMachine`. Examples of these additional resources are:
+* Secrets, required to eventually start the receiver pod.
+* ConfigMaps, required to eventually start the receiver pod.
+* InstanceTypes, required to create the receiving `VirtualMachineInstace`.
+* PreferenceTypes, required to create the receiving `VirtualMachineInstace`.
+
+The source `VirtualMachine` will reference some disks and networks. Whatever creates the target `VirtualMachine` will be responsible for creating blank receiver disks, and properly mapping the networks from the source `VirtualMachine` to the target `VirtualMachine`
+
+When migrating inside the same namespace, the additional `VirtualMachine` and associated resources are not needed as they already exist in the namespace.
+
+The `source` resource will contain a `connectionURL` that points to the synchronization controller at the target. The `source` resource will also contain a key which has to be unique across all migrations. The `target` resource will contain the same key. This will allow the synchronization controller to map the `VirtualMachineInstance` it receives on the synchronization channel to the appropriate `target` `VirtualMachineInstance`. Once the synchronization has been established, any status updates can be communicated between the `source` and the `target` just like in the current state. This means enough information has been exchanged to start the receiver pod, and once the receiver pod is ready, the migration can start.
```

Comment:
why do they need to be in the VMIM? Isn't this information hidden from the user?

Reply:
Can you please add the details what the key is for?

Reply:
> why do they need to be in the VMIM? Isn't this information hidden from the user?
We do not anticipate users interacting with these low level building blocks directly. We have an orchestrator that the users will interact with. The orchestrator will be the one that creates all the resources, including the VMIM. And it can look up the synchronization channel endpoint. This is the one piece of information that needs to be provided to the source, so we can establish the synchronization channel.

> Can you please add the details what the key is for?
On line 153, I give some details. It is just an identifier so we can map the source to the target.

Reply:
Similar comment as for the runStrategy, why in the spec and not in the status? If you put in the spec it means that the key is generated by the overlay orchestra tor, not by KubeVirt. Is this the case?